geopy/geocoders/geocodio.py

```diff
@@ -0,0 +1,243 @@
+            raise ConfigurationError('Must provide an api_key.')
+        self.api_key = api_key
+
+    def geocode(
+        self,
+        query=None,
+        *,
+        limit=None,
+        exactly_one=True,
+        timeout=DEFAULT_SENTINEL,
+        street=None,
+        city=None,
+        state=None,
+        postal_code=None,
+        country=None
+    ):
+        """Return a location point by address. You may either provide a single address
+        string as a ``query`` argument or individual address components using the
+        ``street``, ``city``, ``state``, ``postal_code``, and ``country`` arguments.
+
+        :param str query: The address or query you wish to geocode. You must either
+            provide this argument or a valid combination of ``street``, ``city``,
+            ``state``, and ``postal_code`` and you may not provide those arguments if
+            also providing ``query``.
+
+        :param int limit: The maximum number of matches to return. This will be reset
+            to 1 if ``exactly_one`` is ``True``.
+
+        :param bool exactly_one: Return one result or a list of results, if
+            available.
+
+        :param int timeout: Time, in seconds, to wait for the geocoding service
+            to respond before raising a :class:`geopy.exc.GeocoderTimedOut`
+            exception. Set this only if you wish to override, on this call
+            only, the value set during the geocoder's initialization.
+
+        :param str street: The street address to geocode. If providing this argument
+            you must provide at least one of ``city``, ``state``, or ``postal_code``, and
+            you must *not* provide a ``query`` argument.
+
+        :param str city: The city of the address to geocode. If providing this argument
+            you must *not* provide a ``query`` argument.
+
+        :param str state: The state of the address to geocode. If providing this argument
+            you must *not* provide a ``query`` argument.
+
+        :param str postal_code: The postal code of the address to geocode. If providing
+            this argument you must *not* provide a ``query`` argument.
+
+        :param str country: The country of the address to geocode. If providing this
+            argument you must *not* provide a ``query`` argument.
+
+        :rtype: ``None``, :class:`geopy.location.Location` or a list of them, if
+            ``exactly_one=False``.
+        """
+        if query is not None and \
+                any(p is not None for p in (city, state, postal_code, country)):
+            raise GeocoderQueryError('Address component must not be provided if '
+                                     'query argument is used.')
+        if street is not None and \
+                not any(p is not None for p in (city, state, postal_code)):
+            raise GeocoderQueryError('If street is provided must also provide city, '
+                                     'state, and/or postal_code.')
```

Comment:
Is there a reason why these conditions must be checked on the client-side?

(I'm okay with keeping these unless they might change in future).

Reply:
I think the only argument would be to save making an invalid request, which might count against rate limits or account request limits.